components/Layout.tsx

```diff
@@ -0,0 +1,239 @@
+import { ReactNode, useEffect, useState } from 'react';
+import Link from 'next/link';
+import { useRouter } from 'next/router';
+import { ApolloProvider, useApolloClient } from '@apollo/client/react';
+import { ApolloClient, InMemoryCache } from '@apollo/client';
+import styled from '@emotion/styled';
+import {
+  DirectionalHint,
+  Dropdown,
+  IDropdownStyles,
+  IDropdownOption,
+  IRenderFunction,
+  Icon,
+  NeutralColors,
+  SearchBox,
+  TooltipDelay,
+  TooltipHost,
+} from '@fluentui/react';
+
+import {
+  getEndpointFromQuery,
+  GraphQLEndPoint,
+  GRAPHQL_ENDPOINTS,
+} from 'lib/graphQLEndPoint';
+import {
+  BlockByIndexQuery,
+  BlockByIndexDocument,
+  BlockByHashQuery,
+  BlockByHashDocument,
+  TransactionByIdQuery,
+  TransactionByIdDocument,
+} from 'src/gql/graphql';
+
+import Wrapper from 'components/Wrapper';
+import { CommonPageProps } from 'lib/staticGeneration';
+import useIdFromQuery from 'lib/useIdFromQuery';
+
+export default function Layout({
+  children,
+  staticEndpoint,
+}: CommonPageProps & {
+  children: ReactNode;
+}) {
+  const { isReady, query } = useRouter();
+  const [endpoint, setEndpoint] = useState(staticEndpoint);
+  const [client, setClient] = useState<ApolloClient<object> | null>(null);
+  useEffect(() => {
+    if (!endpoint && isReady) {
+      setEndpoint(getEndpointFromQuery(query));
+    }
+  }, [endpoint, isReady, query]);
```

Comment:
If you want to use this code block in every page (load and update endpoint).
you can consider to use Context

Reply:
Do you mean this one? https://reactjs.org/docs/context.html Thanks for pointing out! I'll see how I might apply it.

Reply:
After the last commit (which I packed the endpoint retrieval logic to useEndpoint() custom hook), it seems like the amount of code that would be reused with using Contexts would be similar, so I'll consider using it in later iterations if it seems like I'll be using more globally available logic.